isbnbot/normalize_isbns.py

```diff
@@ -0,0 +1,137 @@
+        log_file_datetime = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+        log_file = log_dir + '/%s_%s.log' % (job_name, log_file_datetime)
+        file_handler = logging.FileHandler(log_file)
+        file_handler.setLevel(logging.DEBUG)
+        file_handler.setFormatter(log_formatter)
+        self.logger.addHandler(file_handler)
+
+    @staticmethod
+    def isbn_needs_normalization(isbn: str) -> False:
+        """
+        Returns True if the given ISBN is valid and needs to be normalized (hyphens removed, letters capitalized, etc.)
+        Returns False otherwise
+        """
+        if isbnlib.is_isbn10(isbn) or isbnlib.is_isbn13(isbn):
+            normalized_isbn = isbnlib.get_canonical_isbn(isbn)  # get_canonical_isbn returns None if ISBN is invalid
+            return normalized_isbn and normalized_isbn != isbn
+        return False
+
+    def run(self, dump_filepath: str) -> None:
+        """
+        Performs ISBN normalization (removes hyphens and capitalizes letters)
+
+        dump_filepath -- path to *.txt.gz dump containing editions that need to be operated on
+        """
+        if self.dry_run:
+            self.logger.info('dry_run set to TRUE. Script will run, but no data will be modified.')
+
+        header = {'type': 0,
+                  'key': 1,
+                  'revision': 2,
+                  'last_modified': 3,
+                  'JSON': 4}
+        comment = 'normalize ISBN'
+        with gzip.open(dump_filepath, 'rb') as fin:
+            for row in fin:
+                row = row.decode().split('\t')
+                _json = json.loads(row[header['JSON']])
+                if _json['type']['key'] != '/type/edition': continue
+
+                isbns_by_type = dict()
+                if 'isbn_10' in _json:
+                    isbns_by_type['isbn_10'] = _json.get('isbn_10', None)
+                if 'isbn_13' in _json:
+                    isbns_by_type['isbn_13'] = _json.get('isbn_13', None)
+                if not isbns_by_type: continue
+
+                skip_flag = True
+                for _, isbns in isbns_by_type.items():
+                    for isbn in isbns:
+                        if self.isbn_needs_normalization(isbn):
+                            skip_flag = False
+                            break
+                if skip_flag: continue
+
+                olid = _json['key'].split('/')[-1]
+                edition = self.ol.Edition.get(olid)
+                if edition.type['key'] != '/type/edition': continue
+
+                for isbn_type, isbns in isbns_by_type.items():
+                    isbns = getattr(edition, isbn_type, [])
+                    for isbn in isbns:
+                        if not self.isbn_needs_normalization(isbn): continue
+                        normalized_isbn = isbnlib.get_canonical_isbn(isbn)
+                        setattr(edition, isbn_type, [normalized_isbn])  # FIXME, what if an edition has multiple isbns that need normalization?
```

Comment:
Is it potentially destructive to overwrite the ISBN value? Should the script use `append` instead of `setattr`?

Reply:
If `setattr` saves correctly, then it should be ok, but we need to handle the multiple ISBNs case

Reply:
I'd recommend pulling the `save` operation out of the for loops, and instead of `setattr`, using something like `getattr(edition, isbn_type)[index] = normalized_isbn`

Reply:
Actually, instead save the modified isbns in a new array, and use `setattr` at the end (out of the for loop). That was we can also de a dedupe before the save in case normalized the ISBN caused a dupe to appear.